crates/numcodecs-python/src/schema.rs

```diff
@@ -0,0 +1,486 @@
+        }
+
+        if let Some(default) = parameter.default {
+            docs.push_str(", default = `");
+            docs.push_str(&format!("{default}"));
+            docs.push('`');
+        }
+
+        docs.push(')');
+
+        if let Some(info) = &parameter.docs {
+            docs.push_str(": ");
+            docs.push_str(&info.replace('\n', "\n   "));
+        }
+
+        docs.push('\n');
+    }
+
+    if parameters.named.is_empty() {
+        if parameters.additional {
+            docs.push_str("This codec takes *any* parameters.");
+        } else {
+            docs.push_str("This codec does *not* take any parameters.");
+        }
+    } else if parameters.additional {
+        docs.push_str("\nThis codec takes *any* additional parameters.");
+    }
+
+    docs.truncate(docs.trim_end().len());
+
+    Some(docs)
+}
+
+pub fn signature_from_schema(schema: &Schema) -> String {
+    let parameters = parameters_from_schema(schema);
+
+    let mut signature = String::new();
+    signature.push_str("self");
+
+    for parameter in parameters.named {
+        signature.push_str(", ");
+        signature.push_str(parameter.name);
+
+        if let Some(default) = parameter.default {
+            signature.push('=');
+            signature.push_str(&format!("{default}"));
+        } else if !parameter.required {
+            signature.push_str("=None");
+        }
+    }
+
+    if parameters.additional {
+        signature.push_str(", **kwargs");
+    }
+
+    signature
+}
+
+#[allow(clippy::too_many_lines)] // FIXME
+fn parameters_from_schema(schema: &Schema) -> Parameters {
+    if schema.as_bool() == Some(true) {
+        return Parameters {
+            named: Vec::new(),
+            additional: true,
+        };
+    }
+
+    let Some(schema) = schema.as_object() else {
+        return Parameters {
+            named: Vec::new(),
+            additional: false,
+        };
+    };
+
+    let mut parameters = Vec::new();
+
+    let required = match schema.get("required") {
+        Some(Value::Array(required)) => &**required,
+        _ => &[],
+    };
+
+    if let Some(Value::Object(properties)) = schema.get("properties") {
+        for (name, parameter) in properties {
+            parameters.push(Parameter {
+                name,
+                required: required
+                    .iter()
+                    .any(|r| matches!(r, Value::String(n) if n == name)),
+                default: parameter.get("default"),
+                docs: match parameter.get("description") {
+                    Some(Value::String(docs)) => Some(Cow::Borrowed(docs)),
+                    _ => None,
+                },
+            });
+        }
+    }
+
+    let mut additional = !matches!(schema.get("additionalProperties"), Some(Value::Bool(false)));
+
+    if let Some(Value::Array(variants)) = schema.get("oneOf") {
+        additional |= !matches!(schema.get("additionalProperties"), Some(Value::Bool(false)));
```

Comment:
if any enum variant has additional properties, the flattened version has additional properties